### DISCUSSION TITLE
AWS_IO_DNS_QUERY_FAILED

### DISCUSSION THREAD

**Discussion by farzadpanahi** [awslabs/aws-crt-python#502] [General]
I am randomly getting this exception:

```
awscrt.exceptions.AwsCrtError: AWS_IO_DNS_QUERY_FAILED: A query to dns failed to resolve.
```

I am using the latest packages:
```
awscrt==0.19.1
awsiotsdk==1.19.0
boto3==1.28.42
botocore==1.31.42
```

This is the code:

```
        mqtt_connection = mqtt_connection_builder.mtls_from_bytes(
            endpoint=aws_endpoint,
            client_id=client_id,
            cert_bytes=cert_bytes,
            client_bootstrap=client_bootstrap,
            pri_key_bytes=pri_key_bytes,
            ca_bytes=root_ca_bytes,
            clean_session=False,
            on_connection_interrupted=self.__on_mqtt_connection_interrupted,
            on_connection_resumed=self.__on_mqtt_connection_resumed,
        )

        connect_future = mqtt_connection.connect()
        out = connect_future.result()
```
Any idea what the cause of this error is? I cannot trace it back in the code either. I just know that it is coming from `result()` function call.

**Comment by bretambrose** [comment]
DNS resolution happens as the first step of connection establishment and in this case, it failed.  The SDK uses system calls (https://github.com/awslabs/aws-c-io/blob/main/source/posix/host_resolver.c#L34-L46 or https://github.com/awslabs/aws-c-io/blob/main/source/windows/host_resolver.c#L35) to resolve names and that's where your failure is coming from.  

You can enable SDK logging (https://github.com/awslabs/aws-crt-python/blob/main/awscrt/io.py#L28) to get the actual OS error code which may give you more insight into what is happening.
